Update array-manager.md (#9016)
magento/devdocs: src/guides/v2.3/extension-dev-guide/framework/array-manager.md

---
group: php-developer-guide
title: Array Manager

> contributor_name: Christophe Ferreboeuf

Reviewer: Please remove the contributor items as this is not a new topic

Author: I spent 4h checking different scenarios as the documentation in the code itself is really sparse, I refactored most of the article as the examples were mostly copy-paste from existing code and I need to remove the bit of recognition this work can bring me. Sorry but it does not interest me. You can do whatever you want with this PR but I will not remove my contribution items.

|--- |--- |
| `exists` | Checks if the node exists in a given associative array |
| `find` | Finds a node in a nested array and saves its index and parent node reference |
| `findPath` | Get first matching path for elements with specified indexes. It calls PHP `reset` function on `findPaths` results |

> | `findPaths` | Gets matching paths for elements with specified indexes. `$startPath` permits to perform a search starting from specific path. `$internalPath` permits to retrieve only elements that are situated under the specified path. |

Reviewer: ```suggestion
| `findPaths` | Gets matching paths for elements with specified indexes. `$startPath` is used to define a search starting from specific path. `$internalPath` retrieves only elements that are situated under the specified path. |
```

| `set` | Sets the value into a node and returns modified data. |
| `slicePath` | Retrieves a slice of the specified path. |

> Suppose you have the following nested array:

Reviewer: ```suggestion
Take the following nested array as an example :
```

```php
$data = [
    'response' => [
        'status' => 'OK',
        'result' => [
            'items' => [
                0 => 'First result item',
                1 => 'Second result item',
            ],
        ],
        'items' => [
            0 => 'First response item',
            1 => 'Second response item',
        ]
    ],
    'error' => [
        'items' => [
            0 => 'First item',
            1 => 'Second item',
        ]
    ]
];

### Accessing data
```

You can use the  `Magento\Framework\Stdlib\ArrayManager` library to access items in the array:

```php
...

if ($this->arrayManager->get('response/status', $data) === 'OK') {
    $items = $this->arrayManager->get('response/result/items', $data) ?? [];

    foreach ($items as $item) {
        ...
    }
}

...
```

If we use `findPaths` without specifying `$startPath` or `$internalPath`, the function will return all paths containing `items`.

```php
...

$this->arrayManager->findPaths('items', $data);

...
```

```php
[
  0 => "response/items"
  1 => "error/items"
  2 => "response/result/items"
]
```

> If we use `findPaths` with specifying `$startPath`, the function will return paths that contain `items` under the path `response`.

Reviewer: ```suggestion
If we use `findPaths` byspecifying `$startPath`, the function will return paths that contain `items` under the path `response`.
```

```php
...

$this->arrayManager->findPaths('items', $data, 'response');

...
```

```php
[
  0 => "response/items"
  1 => "response/result/items"
]
```

> If we use `findPaths` with specifying `$internalPath`, the function will return paths that contain `items` that are children of element `result`.

Reviewer: ```suggestion
If we use `findPaths` by specifying `$internalPath`, the function will return paths that contain `items` that are children of element `result`.
```

```php
...

$this->arrayManager->findPaths('items', $data, null,'result');

...
```

```php
[
  0 => "response/result/items"
]
```

### Array modification

You can use the  `Magento\Framework\Stdlib\ArrayManager` library to populate an array from the given path if it does not exist yet:

```php
...

$this->arrayManager->populate('response/result/headers', $data)

...
```

> It is possible to modify array in different manners for several values, `merge` is a good option but just one value,

Reviewer: ```suggestion
It is possible to modify an array in different ways for several values, `merge` is one option but for just one value,
```
